Exclude paths from redirect

marksweston opened the pull request:
My team has been using geo_redirect in our project for the last month, which has been working well.  However we need to exclude at least one URL from redirection in a similar way to that discussed in issue #19 

This pull request implements the `:exclude` option with the ability to pass one or more strings representing excluded paths.  Hopefully the addition to the README describes it clearly enough.

n0nick commented on lib/geo_redirect/middleware.rb:
```diff
@@ -60,9 +61,17 @@ def force_redirect?
     end
 
+    def path_excluded?(url)
+      @excludes.reduce(false){ | result, exclude | result || url.path == exclude }
```
1. Why not use [#any?](http://ruby-doc.org/core-2.2.2/Enumerable.html#method-i-any-3F)

2. What do you think about allowing regular expressions here?
So line 18 could be
```ruby
@exclude = Regexp.union(options[:exclude]) if @options[:exclude]
```
and then here in 72 all we need is
```ruby
@exclude && !(@exclude =~ url).nil?
```

n0nick commented on spec/geo_redirect_spec.rb:
```diff
@@ -227,6 +230,42 @@ def should_remember_country(country)
+      end
+
+      context 'when the request URL matches one of the excluded paths' do
+        before {mock_request_from 'US', path: '/exclude_me?query_param=value'}
```
add some spaces inside these braces :) and in line 248 too

n0nick commented:
Thanks for the patch!! Much appreciated.
It looks simple enough. I can merge once the minor code style issues are done, and regarding allowing regular expressions you can either include it yourself or I could merge and add it afterwards, up to you.

marksweston commented:
Updated for code style in the specs.

Updated to use `#any` because yeah, that's obviously so much more intention-revealing than my first attempt.

I'm not so sure about regular expressions though.  When I started working on this patch it was with an assumption that I'd be using a regexp in the logic, but then I began to worry about partial matches.

If I treat the excludes option as a regex, then when I pass the option `exclude: '/exclude'` I would be excluding `/exclude`, `/exclude/foo` and `/exclude/foo/bar`
Less intuitively, I'd also be excluding `/foo/bar/exclude`, `/foo/exclude/bar` and even `/excludes`.
And with this API, a user of the library has no way of limiting these partial matches if they don't want them.

Exact matching of the path component of the URL feels to me like a much safer and more controllable behaviour.

What do you think?

n0nick commented:
Yeah, I guess you're right regarding partial matches. We could solve it by treating strings differently (comparing with `==`), but that would complicate the implementation, probably unnecessarily.
